keps/sig-node/3815-pod-condition-for-configuration-errors/README.md

```diff
@@ -0,0 +1,1021 @@
+### Notes/Constraints/Caveats (Optional)
+
+#### Pending Pod State Review
+
+##### Invalid Image Name
+  - Reproduction: Run image with invalid image name (all capitals)
+  - Pod status:
+    - status: Pending
+  - ContainerStatus:
+    - state=Waiting
+      - message='Failed to apply default image tag "BUSYBOX": couldn''t parse image
+            reference "BUSYBOX": invalid reference format: repository name must be
+            lowercase'
+      - reason=InvalidImageName
+  - Conditions:
+    - Type=Status
+    - Initialized=True
+    - Ready=False
+    - ContainersReady=False
+    - PodScheduled=True
+    - PodReadyToStartContainers=True
+##### ErrImageNeverPull
+  - Reproduction: container with ImagePullPolicy: never but container doesn't exist locally
+  - Pod status:
+    - status: Pending
+  - ContainerStatus:
+    - state=Waiting
+    - reason=ErrImageNeverPull
+  - Conditions:
+    - Type=Status
+    - Initialized=True
+    - Ready=False
+    - ContainersReady=False
+    - PodScheduled=True
+    - PodReadyToStartContainers=True
+- Container with missing config map
+  - Reproduction: Container with missing config map
+  - Pod status:
+    - status: Pending
+  - ContainerStatus:
+    - state=Waiting
+      - message=configmap "%s" not found
+      - reason=CreateContainerConfigError
+  - Conditions:
+    - Type=Status
+    - Initialized=True
+    - Ready=False
+    - ContainersReady=False
+    - PodScheduled=True
+    - PodReadyToStartContainers=True
+##### CreateContainerConfigError - Missing Key Reference
+  - Reproduction: Container with existing config map but invalid key reference
+  - Pod status:
+    - status: Pending
+  - ContainerStatus:
+    - state=Waiting
+      - message=couldn't find key special.how in ConfigMap default/special-config
+      - reason=CreateContainerConfigError
+  - Conditions:
+    - Type=Status
+    - Initialized=True
+    - Ready=False
+    - ContainersReady=False
+    - PodScheduled=True
+    - PodReadyToStartContainers=True
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The impact will be a new condition that gets set to True if a user has a error during startup of their pod.
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+Add a PodFailedToStartup condition in core/v1:
+
+```go
+const (
+  ...
+	// PodFailedToStart For Setting Pod Failed To Start if Pod is stuck due
+	// to configuration error.
+	PodFailedToStart PodConditionType = "FailedToStart"
```

Comment:
All of the above cases are "failing" but not "failed" - ING not ED.  They can all be retried - the image can be updated from invalid->valid, something else on the node might pull the image to satisfy pull policy, the configmap might be created.

We should be clear that this indicates a potentially resolvable situation.

Reply:
Good call out.